qi-test/tests/partition.rkt

```diff
@@ -0,0 +1,82 @@
+#lang racket/base
+
+(provide tests)
+
+(require qi
+         (for-syntax racket/base)
+         rackunit)
+
+;; reference implementation
+(define-qi-syntax-parser partition-inefficient
```

Comment:
Testing a reference implementation is tempting, like validation of testing assumptions prior to doing real testing, but as far as I know it isn't recommended practice, since we're essentially testing the testing apparatus rather than the library code. Here's what I would suggest instead:

1. Add the reference implementation as the initial version of `partition`, along with the unit tests, and commit that
2. Run `make profile-selected-forms` to profile this form specifically
3. Replace the partition implementation with the new one and ensure the tests still pass (so that they are now regression tests), and commit that
4. Re-run the profiler for the form to see the performance change 😎 

P.S. This is exactly where [continuous performance benchmarking](https://github.com/countvajhula/qi/issues/20#issuecomment-1158387769) would tell a nice story, too bad we don't have that yet!

Reply:
I like the idea to run two commits, verifying that one doesn't change the semantics of the other. Will do.